File: leetcode/product_of_array_except_self.cpp
class Solution {
public:
    bool check_zero(vector<int> nums){
        for (int i = 0; i < nums.size(); i++){
            if (nums[i] == 0) return true;
        }
        return false;
    }

    vector<int> productExceptSelf(vector<int>& nums) {
        int total_product = 1;
        int zero_nums = 0;
        for (int i = 0; i < nums.size(); i++){
            if (nums[i] != 0){
                total_product = total_product * nums[i];
            }
            if (nums[i] == 0){
                zero_nums++;
            }
        }

        if (zero_nums == nums.size()) return nums;

        vector<int> result;

        if (nums.size() == 2 && zero_nums == 1){
            for (int i = 0; i < nums.size(); i++){
                result.push_back(nums[1 - i]);
            }
            return result;
        }

        if (zero_nums == nums.size() - 1 || zero_nums >= 2){
            for (int i = 0; i < nums.size(); i++){
                result.push_back(0);
            }
            return result;
        }

        if (check_zero(nums) == true){
            for (int i = 0; i < nums.size(); i++){
                if (nums[i] == 0){
                    result.push_back(total_product);
                }
                else{
                    result.push_back(0);
                }
            }
        }

        else{
            for (int i = 0; i < nums.size(); i++){
                if (nums[i] != 0){
                    result.push_back(total_product / nums[i]);
                }
                else{
                    result.push_back(total_product);
                }    
            }
        }

        return result;
    }
};

File: leetcode/golang/product_of_array_except_self.go
func productExceptSelf(nums []int) []int {
    var result []int
    var zeroes int
    var product int = 1

    for _, value := range nums {
        if value == 0 {
            zeroes += 1
            continue 
        } else {
            product *= value
        }
    }

    for _, value := range nums {
        // if there are 2 or more zeroes, all the elements would be 0
        if zeroes > 1 {
            result = append(result, 0)
        } else if zeroes == 1 {
            if value == 0 {
                result = append(result, product)
            } else {
                result = append(result, 0)
            }
        } else {
            result = append(result, product / value)
        }
    }
    
    return result
}
